src/cmd/services/m3coordinator/downsample/metrics_appender.go

```diff
@@ -363,13 +399,184 @@ func (a *metricsAppender) debugLogMatch(str string, opts debugLogMatchOptions) {
+	// Take an encoder from the cached encoder list, if not present get one
+	// from the pool. Add the returned encoder to the used list.
+	var tagEncoder serialize.TagEncoder
+	if len(a.cachedEncoders) == 0 {
+		tagEncoder = a.tagEncoderPool.Get()
+	} else {
+		tagEncoder = a.cachedEncoders[0]
+		a.cachedEncoders = a.cachedEncoders[1:]
+	}
+	a.inuseEncoders = append(a.inuseEncoders, tagEncoder)
+	tagEncoder.Reset()
+	return tagEncoder
+}
+
+func (a *metricsAppender) newSamplesAppenders(
+	stagedMetadata metadata.StagedMetadata,
+	unownedID []byte,
+) ([]samplesAppender, error) {
+	// Check if any of the pipelines have tags or a graphite prefix to set.
+	var tagsExist bool
+	for _, pipeline := range stagedMetadata.Pipelines {
+		if len(pipeline.Tags) != 0 || len(pipeline.GraphitePrefix) > 0 {
+			tagsExist = true
+			break
+		}
+	}
+
+	// If we do not need to do any tag augmentation then just return.
+	if !a.augmentM3Tags && !tagsExist {
+		return []samplesAppender{
+			{
+				agg:             a.agg,
+				clientRemote:    a.clientRemote,
+				unownedID:       unownedID,
+				stagedMetadatas: []metadata.StagedMetadata{stagedMetadata},
+			},
+		}, nil
+	}
+
+	var (
+		pipelines []metadata.PipelineMetadata
+		appenders []samplesAppender
+	)
+	for _, pipeline := range stagedMetadata.Pipelines {
+		// For pipeline which have tags to augment we generate and send
+		// separate IDs. Other pipelines return the same.
+		pipeline := pipeline
+		if len(pipeline.Tags) == 0 && len(pipeline.GraphitePrefix) == 0 {
+			pipelines = append(pipelines, pipeline)
+			continue
+		}
+
+		tags := augmentTags(a.tags, pipeline.GraphitePrefix, pipeline.Tags, pipeline.AggregationID)
+
+		sm := stagedMetadata
+		sm.Pipelines = []metadata.PipelineMetadata{pipeline}
+
+		appender, err := a.newSamplesAppender(tags, sm)
+		if err != nil {
+			return []samplesAppender{}, err
+		}
+		appenders = append(appenders, appender)
+	}
+
+	if len(pipelines) == 0 {
+		return appenders, nil
+	}
+
+	sm := stagedMetadata
+	sm.Pipelines = pipelines
+
+	appender, err := a.newSamplesAppender(a.tags, sm)
+	if err != nil {
+		return []samplesAppender{}, err
+	}
+	return append(appenders, appender), nil
+}
+
+func (a *metricsAppender) newSamplesAppender(
+	tags *tags,
+	sm metadata.StagedMetadata,
+) (samplesAppender, error) {
+	tagEncoder := a.tagEncoder()
+	if err := tagEncoder.Encode(tags); err != nil {
+		return samplesAppender{}, err
+	}
+	data, ok := tagEncoder.Data()
+	if !ok {
+		return samplesAppender{}, fmt.Errorf("unable to encode tags: names=%v, values=%v", tags.names, tags.values)
+	}
+	return samplesAppender{
+		agg:             a.agg,
+		clientRemote:    a.clientRemote,
+		unownedID:       data.Bytes(),
+		stagedMetadatas: []metadata.StagedMetadata{sm},
+	}, nil
+}
+
+func augmentTags(
+	originalTags *tags,
+	graphitePrefix [][]byte,
+	t []models.Tag,
+	id aggregation.ID,
+) *tags {
+	// Create the prefix tags if any.
+	tags := newTags()
```

Comment:
This seems like another "allocation per thing processed", can we use the same pattern here with "cachedEncoders" and "inuseEncoders" to avoid that? It's fine to call "newTags()" I think but amortizing when that happens so that in the steady state it doesn't need to do that would be ideal.

Reply:
sure, yeah.. that should help a bit at least..